test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp

```diff
@@ -0,0 +1,517 @@
+        "GetFieldName failed");
+    if (is_static_field(env, klass, fid)) {
+        verify_int_field(env, klass, fid);
+        value = env->GetStaticIntField(klass, fid);
+    } else {
+        value = 0;
+    }
+}
+
+void Klass::explore_fields(JNIEnv* env) {
+    jint this_count;
+    jfieldID* this_fields;
+    check_jvmti_status(env,
+        jvmti->GetClassFields(klass, &this_count, &this_fields),
+        "GetClassFields failed");
+
+    jint super_count = super_klass != nullptr ? super_klass->field_count : 0;
+
+    fields = (Field*)allocate(env, sizeof(Field) * (super_count + this_count));
+    field_count = 0;
+
+    if (super_klass != 0) {
+        // super_klass->fields already contains fields from all superclasses in the required order.
+        for (int i = 0; i < super_count; i++) {
+            fields[field_count++].init(env, super_klass->klass, super_klass->fields[i].id);
+        }
+    }
+
+    // Add field of this class to the end of the list.
+    for (int i = 0; i < this_count; i++) {
+        fields[field_count++].init(env, klass, this_fields[i]);
+    }
+    deallocate(env, this_fields);
+}
+
+
+// Calculates maximum number of implemented interfaces of the klass and its superinterfaces.
+static jint get_max_interface_count(JNIEnv* env, jclass klass) {
+    jint interface_count;
+    jclass* interfaces;
+    check_jvmti_status(env,
+        jvmti->GetImplementedInterfaces(klass, &interface_count, &interfaces),
+        "GetImplementedInterfaces failed");
+
+    jint result = interface_count;
+    // interfaces implemented by superinterfaces
+    for (jint i = 0; i < interface_count; i++) {
+        result += get_max_interface_count(env, interfaces[i]);
+    }
+
+    deallocate(env, interfaces);
+
+    return result;
+}
+
+// Explores all interfaces implemented by 'klass' sorting out duplicates
+// and store them in the 'arr' starting from 'index'.
+// Returns number of the interfaces added.
+static jint fill_interfaces(Klass** arr, jint index, JNIEnv* env, jclass klass) {
+    jint interface_count;
+    jclass* interfaces;
+    check_jvmti_status(env,
+        jvmti->GetImplementedInterfaces(klass, &interface_count, &interfaces),
+        "GetImplementedInterfaces failed");
+
+    jint count = 0;
+    for (jint i = 0; i < interface_count; i++) {
+        // Skip interface if it's already in the array
+        // (i.e. implemented by another superclass/superinterface).
+        bool dup = false;
+        for (jint j = 0; j < index; j++) {
+            if (env->IsSameObject(arr[j]->klass, interfaces[i]) == JNI_TRUE) {
+                dup = true;
+                break;
+            }
+        }
+        if (dup) {
+            continue;
+        }
+
+        // Add the interface.
+        arr[index + count] = Klass::explore(env, interfaces[i]);
+        count++;
+
+        // And explore its superinterfaces.
+        count += fill_interfaces(arr, index + count, env, interfaces[i]);
```

Comment:
I assume here we are always dealing with test classes that are known not to be deeply nested.

Reply:
Yes, we explore only test classes/interfaces